Level-3 HK database and loading framework

jlashner opened the pull request:
This is a draft PR with an initial level-3 housekeeping indexing and loading framework.

Docs will come shortly, but this can currently be used and tested on satp1 data, since I have a satp1 index that should be up to date as of 03/28.

First create a hk.yaml file to use as configuration, such as:
```
hk_root: /so/data/satp1/hk
hk_db: /so/home/jlashner/work/daily/2024-03-28/l3-hk/satp1_hk.db
aliases:
    fp_temp: cryo-ls372-lsa21yc.temperatures.Channel_02_T
```
Then in python, you can load data the last week's worth of focal-plane temps as follows
```
t1 = time.time()
t0 = t1 - 7*24*3600
cfg = hkdb.HkConfig.from_yaml('hk.yaml')
ls = hkdb.LoadSpec(
    cfg=cfg,
    fields=['cryo-ls372-lsa21yc.*.*'],
    start=t0, end=t1,
)
res = hkdb.load_hk(ls, show_pb=True)
plt.plot(*res.fp_temp)

# The above is the same as below, attributes in the result are set based on aliases.
# plt.plot(*res.data['cryo-ls372-lsa21yc.temperatures.Channel_02_T'])
```

kmharrington commented on sotodlib/io/hkdb.py:
```diff
@@ -0,0 +1,366 @@
+        self.fields = fs
+
+
+class HkResult:
```
can this just be an axis manager?

jlashner replied:
I don't think this quite fits into an AxisManager so easily... because hk fields aren't cosampled, most of the benefits of AxisManager aren't really applicable, and I'm not sure how axis-managers handle long field names that aren't proper python identifiers, such as `cryo-ls372-lsa21yc.temperatures.Channel_02_T`.

I think it would be better to leave this as is, and have a separate function that transforms this into an AxisManager when re-sampling hk to be aligned with the 'samps' axis.

kmharrington replied:
Sanah did the setup for getting axismanagers where the fields are co-sampled and then co-sampling. I don't see a reason not to use it?

jlashner replied:
I really dislike the idea of making this load fn dependent on all of the hk -> axis-manager conversion stuff that is in hk_utils...

I think there is enough subtlety in how fields are grouped and how the cosampling is done, that it should happen separately and not as a part of the loading function.

kmharrington commented:
SQLAlchemy has a lot written up on how to use sessions. The starting them inside of functions like what's done here isn't recommended. I'm having multiple issues with that setup inside of the rest of data-packaging.  [See the thingOne, thingTwo "don't do this" section.]( https://docs.sqlalchemy.org/en/20/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it ) The imprinter/other HKdb setup is closer to what they recommend (although not completely there). At least some of the g3tsmurf issues comes from the rollbacks instead of throwing errors and some of the readout_id querying generating their own sessions.
 
For that reason, I'm really not of fan of making these "functions passing around a database that sometimes has a session" instead of a "class with a stable session." But even if it stays functions instead of classes:  we don't want that setup in `index_all` where that function begins a session and then calls a function that begins a session. At an absolute minimum, nesting functions that make sessions is going to be quite confusing. Also, [I'm pretty sure that a transaction is issued every time we start a session](https://docs.sqlalchemy.org/en/20/orm/session_transaction.html) so that's going to hit the NFS mount way more than alternative setups.

jlashner commented:
Thanks katie for pointing me to those sqlalchemy pages, I think I addressed your concern by allowing for passing in of a session from the outer function, to avoid nesting.

I'm going to open this up as being "ready for review" since I've added docs, and I have satp1 and satp3 dbs accessible here:
```
- /so/home/jlashner/work/daily/2024-03-28/l3-hk/satp1_hk.db
- /so/home/jlashner/work/daily/2024-03-28/l3-hk/satp3_hk.db
```

sanahabhimani commented:
Testing this in prep to integrate into `hk_utils`--- I was playing around with `LoadSpec`. My config file has 2 fields/aliases I input. Currently, if I want to load data only from the fields in my config file, I set `fields = []`. Is it worth having `fields=None` as a default argument for  a case like this?

jlashner commented:
@kmharrington and @mhasself I think I addressed all of your comments from the data-packaging call last week, by:

- adding size, mod_time, index_status to HkFiles table
- adding lookback time option to the cfg object to limit scanning entire hk archive
- adding caching of failed files (through the index_status)

I also added saving / loading of the HkResults object because I find that really useful.

I don't think this needs a separate script for updating, since I believe prefect can just use the `update_index_all` as an entrypoint, but I'll check with Yulin.

@sanahabhimani I do get where you're coming from, but when building and testing I was sort of envisioning the hkcfg files having many aliases available, and for it being kind of rare to need to load all of them at once. So I'm not too keen on making this the default behavior, if that makes sense. Thanks for the suggestion though!

sanahabhimani commented:
I can push updates to hk_utils with this new framework to this branch if that's the cleanest way to go.

mhasself commented on sotodlib/io/hkdb.py:
```diff
@@ -0,0 +1,425 @@
+            cfg = HkConfig.from_yaml(cfg)
+        self.cfg = cfg
+
+        self.engine = db.create_engine(f"sqlite:///{cfg.hk_db}", echo=cfg.echo_db)
```
You could generalize this, for non-sqlite options, by accepting engine as a kwarg.

That would facilitate experimentation with alternative databases.

jlashner replied:
This sounds good to me, but instead of passing through an engine I'd rather have the `db_url` settable via the cfg, either as a string or as a [sqlalchemy.url](https://docs.sqlalchemy.org/en/20/core/engines.html#creating-urls-programmatically) object, so it is more straightforward to set via the cfg file, and we don't have to worry about passthrough args. Are there any reasons you think its better for the engine object to be passed in over the URL?

mhasself replied:
Oh, ya, that sounds fine.

mhasself commented on sotodlib/io/hkdb.py:
```diff
@@ -0,0 +1,425 @@
+    new_files = sorted(list(set(all_files) - set(existing_files)))
+
+    files = []
+    print(f"Adding {len(new_files)} new files to index...")
```
Use logger instead of print for library-level code.

mhasself commented on sotodlib/io/hkdb.py:
```diff
@@ -0,0 +1,425 @@
+    For example::
+
+        {
+            'fp_temp': 'cryo-ls372-lsa21yc.temperatures.Channel_02_T',
```
Perhaps warn the user that aliases should be valid python attribute names, since you setattr with these, later.  (I.e. they should not use 'fp-temp'.)

mhasself commented on sotodlib/io/hkdb.py:
```diff
@@ -0,0 +1,425 @@
+        return cls(data, aliases=aliases)
+
+
+def load_hk(load_spec: LoadSpec, show_pb=False):
```
Since this is a main entry point, it might be helpful to also accept "load_spec" as a dict that gets auto-converted to LoadSpec.  Also consider engine=... arg to pass through to HkDb.

mhasself commented on docs/hkdb.rst:
```diff
@@ -0,0 +1,102 @@
+    aliases:
+        fp_temp: cryo-ls372-lsa21yc.temperatures.Channel_02_T
+
+In addition to the hk-root and database paths, an "alias" dictionary can
```
Mention that aliases are pertinent to _load_ only, and don't affect how info is stored in the database.

Might be helpful to have an example where a common database configuration is loaded from file, but user modifies (in script) the aliases block to suit their current purpose.  (I.e. emphasize that aliases are a dynamic element to tune for each use case, whereas a database instance is a general storage tool useful for many purposes.)

mhasself reviewed (approved):
Minor suggestions, but I won't insist.

I get that this is sort of a "minimum working implementation", but there are lots of obvious extensions / conveniences one might pursue (e.g. more command line args for update_hkdb.py).

Also -- site_pipeline.rst should an entry for update-hkdb.

jlashner commented:
Made a few changes addressing Matthew's suggestions. In particular, expanded HkConfig so that it supports generic database URLs, so it can support more than sqlite databases.

This is currently tested and running from a postgres db on storage1.
